### DISCUSSION TITLE
[FeatureRequest] Game scaling

### DISCUSSION THREAD

**Discussion by crypt47** [ihhub/fheroes2#7046] [Ideas]
Pixels are getting smaller and smaller and for me it's a deal breaker to play HHoM2 on modern monitors. Lots of games already implemented pixel scaling by nearest neighbor, so will fheroes2 I hope.

**Comment by Branikolog** [comment]
Hi, @crypt47 !
Nearest scaling can be switched on in .cfg file. Is this what you're looking for?

**Reply by crypt47** [reply]
Hi, I don't know yet. The last time I saw the game there was nothing. Do you have a documentation for .cfg file?

**Reply by ihhub** [reply]
Just open the configuration file. Each option has a comment.

**Reply by crypt47** [reply]
> Just open the configuration file. Each option has a comment.

it's not that easy. on FreeBSD I have it distributed like this:

```
fheroes2-1.0.0:
	/usr/local/bin/fheroes2
	/usr/local/share/applications/fheroes2.desktop
	/usr/local/share/fheroes2/files/data/resurrection.h2d
	/usr/local/share/icons/hicolor/128x128/apps/fheroes2.png
	/usr/local/share/metainfo/fheroes2.appdata.xml
```

No configs, no man pages... So I have to go somewhere and take something for an example...

**Reply by crypt47** [reply]
Ok, it creates the .fheroes2/fheroes2.cfg automatically. Indeed in a new version it has

```
# scaling type: nearest or linear (set by default)
screen scaling type = linear
```
So? Where is a scaling factor? I'm talking about scaling sprites and everything. This comment doesn't explain anything to me.

**Reply by ihhub** [reply]
All resolutions are within the game's options. Also hit F4 for full screen.

**Reply by crypt47** [reply]
> All resolutions are within the game's options. Also hit F4 for full screen.

Changing resolution, doesn't change neither the interface elements size, nor fonts... and I suspect nor sprites during the game.

I see no signs of _scalling_ actually, except for the main menu. What I'm talking about is running the game with 800x600 in game resolution and 1600x1200 actual resolution. (Which means scalling factor of 2.)

Some random image from Internet to illustrate: https://forums.rpgmakerweb.com/data/attachments/135/135974-78970e91b0e3843740369dc1268d59ec.jpg

**Reply by ihhub** [reply]
Integer scaling was implemented in 1.0.3: https://github.com/ihhub/fheroes2/pull/6516

**Reply by ihhub** [reply]
Also we aren't responsible for FreeBSD package. Man page should be on shoulders on the package maintainer. If the package is not 1.0.3 there is nothing we can do. Alternatively, you can compile the application on your machine.

**Reply by crypt47** [reply]
Ok, now it's something to test! It wasn't mentioned among new features where I've read about release. And it also didn't get into https://github.com/ihhub/fheroes2/wiki/Features-and-enhancements-of-the-project

**Reply by ihhub** [reply]
This wiki page should be updated and some new features were added.

**Reply by crypt47** [reply]
> This wiki page should be updated and some new features were added.

I've updated to 1.0.3. and as it's undocumented yet, I guess putting this into config should do the trick?

```
videomode = 640x480x2
```
For some reason it doesn't change anything. And it doesn't allow to choose any x2 resolutions at first start. Weird implementation.:(

**Reply by ihhub** [reply]
Are you sure that you use 1.0.3 version? Based on your logs it's 1.0.0.

**Reply by crypt47** [reply]
> Are you sure that you use 1.0.3 version? Based on your logs it's 1.0.0.

```
$ fheroes2 --version
fheroes2 engine, version: 1.0.3
```

1.0.3 it is now...

**Reply by oleg-derevenetz** [reply]
> For some reason it doesn't change anything. And it doesn't allow to choose any x2 resolutions at first start. Weird implementation.:(

How does the resolution selection window look like for you? This is how it looks like for me:

![resolutions](https://user-images.githubusercontent.com/32623900/233738030-6c3db542-be68-47fa-ac5b-fe568d9aab4f.jpg)

Also please note that such scaling is available only with SDL2. I do not know how exactly fheroes2 is built in your case, since the legacy SDL1 is also still supported.

**Reply by crypt47** [reply]
> How does the resolution selection window look like for you? This is how it looks like for me:

https://i.ibb.co/G7QjvbY/2023-04-22-134106-640x480-scrot.png

SDLv2 alright.

**Reply by ihhub** [reply]
@crypt47 , please use an official version of Heroes of Might and Magic 2. You can either purchase from GOG or Ubisoft Store.

**Reply by crypt47** [reply]
> @crypt47 , please use an official version of Heroes of Might and Magic 2. You can either purchase from GOG or Ubisoft Store.

Absolutely, but for now for the sake of test I used whatever I manage to setup last time. Without clear manual to put everything where it belongs was quest by itself.

Shouldn't break the subject, right?

**Reply by crypt47** [reply]
> How does the resolution selection window look like for you? This is how it looks like for me:
> 

I also don't clearly understand what you've mean in #6516.

> you should have 1600x1200 in the list of supported resolutions

What are the 'supported resoultions'? Does the game play cleaver and get the list from OS? Does it obey the config file anyway? Maybe there is a way to force it, if detection is wrong?

**Reply by ihhub** [reply]
We have no control how SDL gets a list of resolutions for FreeBSD. A user experienced similar issues having only a single resolutions returned by SDL so I assume that this is 1920x1280. The rest of resolutions are added automatically by the engine. There is no 1280 x 960 resolution in your list so no integer scaling can be applied. As I mentioned earlier just choose a small resolution and use full screen.

**Reply by crypt47** [reply]
> As I mentioned earlier just choose a small resolution and use full screen.

I don't want to use full screen. I want it to run with scaled resolution as suggested by #6516. So you say it's SDL problem? Ok, I will look into it. But why then can't I run it 640x480 (the engine automatically added, right?) and the engine doesn't automatically added 640x480x2 (scaled)?

**Reply by oleg-derevenetz** [reply]
> But why then can't I run it 640x480 (the engine automatically added, right?) and the engine doesn't automatically added 640x480x2 (scaled)?

Because currently missing resolutions are added only if they are lower than the lowest supported resolution returned by the OS. This is in fact the emergency mechanism designed to deal with resolution detection issues on certain platforms where OS returned just one large resolution to SDL. If the smallest resolution returned from OS to SDL is less than 1280x960, then neither 1280x960 nor (respectively) 640x480x2 will be added.

**Reply by crypt47** [reply]
> Because currently missing resolutions are added only if they are lower than the lowest supported resolution returned by the OS. 

Let's see... I guess, currently reported resolutions by OS can be considered the following:
```
 $ xrandr -q
Screen 0: minimum 8 x 8, current 1920 x 1080, maximum 16384 x 16384
VGA-0 disconnected primary (normal left inverted right x axis y axis)
DVI-D-0 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 475mm x 267mm
   1920x1080     60.00*+
   1280x1024     75.02    60.02  
   1152x864      75.00  
   1024x768      75.03    60.00  
   800x600       75.00    60.32  
   640x480       75.00    59.94  
HDMI-0 disconnected (normal left inverted right x axis y axis)
DVI-D-1 disconnected (normal left inverted right x axis y axis)

```

**Reply by oleg-derevenetz** [reply]
> Let's see... I guess, currently reported resolutions by OS can be considered the following:

So the smallest resolution reported by the OS is 640x480, so 1280x960 (and corresponding scaled version 640x480x2) is not added.

**Reply by crypt47** [reply]
> This is in fact the emergency mechanism designed to deal with resolution detection issues on certain platforms where OS returned just one large resolution to SDL. If the smallest resolution returned from OS to SDL is less than 1280x960, then neither 1280x960 nor (respectively) 640x480x2 will be added.

You mean "If the smallest resolution returned from OS to SDL is *bigger* than 1280x960, then neither 1280x960 nor (respectively) 640x480x2 will be added." ??

Probably I need some time to understand.

**Reply by oleg-derevenetz** [reply]
This mechanism was designed to fix a specific issue when there is just one supported resolution (usually the big one like 1980x1020) reported by the OS. In this case adding lower resolutions (1280x960, 800x600, 640x480, etc) is relatively safe action (but still not 100% safe). However, adding resolutions that are larger than that single resolution reported by the OS is dangerous because hardware may not support them.

**Reply by crypt47** [reply]
Hard to understand. Let me start from the beginning...

> This is in fact the emergency mechanism designed to deal with resolution detection issues on certain platforms where OS returned just one large resolution to SDL.

OK, mine doesn't return one, so it's not an emergency then? So everything should be normal, right?

> Because currently missing resolutions are added only if they are lower than the lowest supported resolution returned by the OS.

Why?? I just want a double sized window. Why all this fuss and complications? What the minimal resolution has to do with it? I want the window to be bigger, not smaller! And it should fit on my screen... I don't get it.

>  If the smallest resolution returned from OS to SDL is less than 1280x960, then neither 1280x960 nor (respectively) 640x480x2 will be added.

That's just illogical. I have screen bigger then 1280x960. I want to use the option. Why not?

**Reply by oleg-derevenetz** [reply]
> Why?? I just want a double sized window. Why all this fuss and complications?

Because then you press F4 with this "custom" resolution, get a crash because you video driver does not support the corresponding aspect ratio in the full screen mode and then you file a bug :)

**Reply by crypt47** [reply]
> Because then you press F4 with this "custom" resolution, get a crash because you video driver does not support the corresponding aspect ratio in the full screen mode and then you file a bug :)

Well, ok... Nobody likes to file bugs indeed. But why don't you handle these two situations (windowed/full) separately then...

**Reply by oleg-derevenetz** [reply]
> But why don't you handle these two situations (windowed/full) separately then...

In this case users will be able to use F4 with some resolutions but not others, which is confusing (see "file a bug"). But in fact this may be mitigated using the reverse logic: do not search for bigger resolutions for existing smaller ones (640x480 -> 1280x960) but add lower "pseudo-resolutions" for existing bigger resolutions instead (1280x1024 -> 640x512x2). @ihhub what do you think?

**Reply by crypt47** [reply]
> > But why don't you handle these two situations (windowed/full) separately then...
> 
> In this case users will be able to use F4 with some resolutions but not others, which is confusing (see "file a bug").

This is still very confusing, but I know one thing. When I press F4, I want it to be on full screen with the same pixel (scale) size. For that you have to switch game resolution to nearest suitable and it's ok to have some black borders, if it doesn't fit perfectly on wide screen.

**Reply by crypt47** [reply]
Now I'm searching the way to rid off those resolutions, reported by xrandr, so only max 1920x1080 remained.

**Reply by oleg-derevenetz** [reply]
> For that you have to switch game resolution to nearest suitable and it's ok to have some black borders, if it doesn't fit perfectly on wide screen.

Currently we do not draw any black borders of this kind in the engine. When you have a game window with resolution 1280x960 and try to switch to the fullscreen mode on device that supports only 1280x1024, then the video driver may emulate 1280x960 using black borders, but it also may crash instead, so relying on driver is not safe in this aspect.

**Reply by crypt47** [reply]
> > For that you have to switch game resolution to nearest suitable and it's ok to have some black borders, if it doesn't fit perfectly on wide screen.
> 
> Currently we do not draw any black borders 

What is was trying to say is I would expect a foolproof switching to full screen mode, instead of it plays cleaver and hides something from me (and I file a bug report anyway). Because at this point the feature doesn't work for me at all. I've check Xorg, it gets modes from EDID, so it's normal. What you are doing, when trying to match window and fullscreen resolution is not.

Btw how to call a resolution change screen again after first run?

**Reply by oleg-derevenetz** [reply]
> What is was trying to say is I would expect a foolproof switching to full screen mode

Yeah, that's why it does not currently add unsupported resolutions left and right.

> Btw how to call a resolution change screen again after first run?

New Game (on main screen) -> Settings -> Graphics Settings.

**Reply by crypt47** [reply]
> Yeah, that's why it does not currently add unsupported resolutions left and right.
>

It would be better if it change resolution correctly at full screen according to max hardware resolution...
 
> New Game (on main screen) -> Settings -> Graphics Settings.

Thanks. I couldn't rid off system resolutions with xrandr (and I don't consider this hacking as valid way to run a game), so I patched your source so now it adds ''unsupported resolutions" with the rest of others and I finally could use them and do something more fun with the program.

**Reply by oleg-derevenetz** [reply]
> It would be better if it change resolution correctly at full screen according to max hardware resolution...

It does change resolution "correctly". What you are in fact asking is to support non-standard resolutions (not supported by the hardware) in the fullscreen mode by using a different resolution and customizing the in-game picture with "black bars". Sorry, but this is not a priority.

As for the logic of finding the available scaled resolutions, I proposed the PR #7051 with the logic described [here](https://github.com/ihhub/fheroes2/discussions/7046#discussioncomment-5695705) - there should be 960x540x2 "virtual" resolution for your 1920x1080 "hardware" resolution and 640x512x2 "virtual" for your 1280x1024 "hardware", both are safe for the fullscreen mode.

As for windowed mode, with SDL2 you can do this:

https://user-images.githubusercontent.com/32623900/233808803-7a89a845-3585-4ca2-b836-89d3f29e73e0.mp4

but the window size is currently not saved to the `fheroes2.cfg`.

**Reply by crypt47** [reply]
> What you are in fact asking is to support non-standard resolutions

I don't see why. I ask you to put FullHD picture on my FullHD monitor and the in-game resolution stays 640x480. Anyway see below, how I done that.

> but the window size is currently not saved to the `fheroes2.cfg`.

I played windowed mode yesterday. What I did:

1) Added x2 modes in your code.
2) Selected 640x480x2.
3) Undecorated and Maximized window by means of WM, not F4. Some open source GUI can save such properties per application.

As a result not only I got what I need, the mode also doesn't have black stripes on top and bottom. Vertical size of the picture is 1080. So yes, there comes "non-standard resolutions". Even though I wasn't expecting to get it. Maybe you can explain. Also please notice, that deviation from your logic and changing your code gives perfect results.

The cursor doesn't scale, but that's a minor issue. 

![2023-04-23-005937_1920x1080_scrot](https://user-images.githubusercontent.com/45311771/233829502-8cf34d5c-fd7e-4eeb-9f86-a1eb1783c621.png)

**Comment by oleg-derevenetz** [comment]
> I don't see why. I ask you to put FullHD picture on my FullHD monitor and the in-game resolution stays 640x480.

Once again. You can do this in windowed mode by stretching the window. But scaled resolution 640x480x2 requires to have support for the corresponding "native" resolution 1280x960, because when you press the F4 to enter the full-screen mode, the video driver will be requested to switch to 1280x960 and it will fail to do that, so we can't just add such resolutions to the list left and right. However, when you press F4 with the stretched window with 640x480 resolution (even if it is stretched to 1920x1080, just like in your case - in fact, it's 1440x1080 with black bars), app will still ask the video driver for 640x480 and if it supported (and it is supported in your case according to the `xrandr` output), then no issues will arise.

> Anyway see below, how I done that.
> Maximized window by means of WM, not F4.

So you did it in the windowed mode, not in the full screen mode. No video mode change was performed by the video driver.

> As a result not only I got what I need, the mode also doesn't have black stripes on top and bottom.

If your current "native" resolution is 1920x1080, then 640x480 will be scaled to 1440x1080, so you will have the black stripes left and right, not top and bottom.

> Also please notice, that deviation from your logic and changing your code gives perfect results.

Only while you stay in windowed mode and don't ask the video driver to switch to this "non-standard resolution" by pressing the F4 button.

> The cursor doesn't scale, but that's a minor issue.

Turning on the software cursor in the `fheroes2.cfg` should resolve this issue.

**Reply by crypt47** [reply]
> Only while you stay in windowed mode and don't ask the video driver to switch to this "non-standard resolution" by pressing the F4 button.

This is whole point I was waiting for scaled resolution actually. To keep native LCD resolution and avoid smearing of the picture. That's why I stated from the beginning I'm not going to use F4. I see very little reasons to dance around native video modes (no 3D, not performance issues). Changing actual screen resolution in HoMM2 is a relic of the past. Nobody really needs game to work in 640x480 on LCD panel. It's a hardware emulation of resolution on LCD even if video driver involved. It's a compatibility mode, not something you have to keep around so carefully. Going with a native physical resolution and scaling the game - seems reasonable.

**Reply by oleg-derevenetz** [reply]
> That's why I stated from the beginning I'm not going to use F4.

**You** are not going :) But an ordinary user may expect that chosen resolution will work everywhere, both in windowed and fullscreen mode, so we need to dance. For example, when I switch to a non-standard (manually added) resolution on my Windows machine and then press F4, no switch to fullscreen is actually performed (however `SDL_SetWindowFullscreen()` call returns no error), but the in-window rendering is frozen until I press F4 once again.

> Going with a native physical resolution and scaling the game - seems reasonable.

SDL2 is able to do this using the special `SDL_WINDOW_FULLSCREEN_DESKTOP` mode, but it doesn't work equally reliably everywhere, so we use it on some platforms but not others.

**Reply by crypt47** [reply]
> But an ordinary user may expect that chosen resolution will work everywhere, both in windowed and fullscreen mode

The ordinary user expects the picture, not correctness of term 'resolution'. If he chooses 640x480 and you go full screen with the native resolution and scale the game, he will be more pleased because NN scaling is used. You just need to scale the game according to physical resolution. No crashing, no bug reporting. I think even if you only put the game in windowed mode by F4, he will be more pleased. Because no matter how you put it, emulation of 640x480 on modern panels looks ugly. I just can't imaging 'ordinary user' who likes this. And that's why the remakes of old games on Android I know do exactly this. They scale.

**Reply by oleg-derevenetz** [reply]
> emulation of 640x480 on modern panels looks ugly

It depends on the size of that panel. For fheroes2, 640x480 in-game resolution on Android phones is in fact the only resolution suitable for comfortable play, and it is used by default (as of 1.0.3 release, but in `master` this is no longer quite the case).

> And that's why the remakes of old games on Android I know do exactly this.

Well, I'm not sure if Android even allows its video driver to satisfy any direct screen resolution change requests from an app, so there's just no other way than scaling.

**Reply by crypt47** [reply]
> > emulation of 640x480 on modern panels looks ugly
> 
> It depends on the size of that panel. 

I'm not talking about in-game resolution here. In-hardware emulation when you put video driver in 640x480 mode is ugly everywhere by definition. That's why traditional F4 function is absolated imho.

> Well, I'm not sure if Android even allows its video driver to satisfy any direct screen resolution change requests from an app, so there's just no other way than scaling.

And that's exactly as it should to be! One native resolution. The ordinary user doesn't need compat mode from 90s! He needs the modern app to work crisply.

**Reply by oleg-derevenetz** [reply]
> In-hardware emulation when you put video driver in 640x480 mode is ugly everywhere by definition.

Well, as far as I'm aware of this topic, often the scaling is performed not even by video driver / video card, but by the monitor itself, and that naturally gives deplorable results, because monitors are relatively stupid. However, some video cards have support for GPU scaling (for example, this can be turned on for some AMD cards), and in this case GPU may (in theory) apply more sophisticated scaling algorithms than those that are currently supported by SDL (linear and nearest neighbor), and 640x480 in "true" full-screen may look better than those that are scaled by software (and work faster because they are performed on GPU, possibly using some this-GPU-specific implementation).

**Reply by crypt47** [reply]
In-hardware - I meant by the monitor. Anyway, I spent sometime with an engine, everything pretty much works, but for some reason the animation blocks any actions of the player. For example: the new structure in the castle takes a while to appear, and you can't just leave. It's everywhere in the game and you can't play it as fast as you like.

**Reply by oleg-derevenetz** [reply]
> For example: the new structure in the castle takes a while to appear, and you can't just leave. It's everywhere in the game and you can't play it as fast as you like.

That's exactly how it works in the original game. In the original game, you should wait until the new building is fully "built" before taking any other action, etc.

**Reply by crypt47** [reply]
> > For example: the new structure in the castle takes a while to appear, and you can't just leave. It's everywhere in the game and you can't play it as fast as you like.
> 
> That's exactly how it works in the original game. In the original game, you should wait until the new building is fully "built" before taking any other action, etc.

I thought so. But it's annoying a bit.

**Reply by oleg-derevenetz** [reply]
> I thought so. But it's annoying a bit.

Well, there is, for example, a proposal to give the player the opportunity not to wait for the end of the introductory music at the beginning of the battle: #7067. But it is not merged yet. Perhaps in the future there will be similar proposals for other pauses of this kind.
